File: Fundamentals_of_Software_Design/2_Algorithms/Поиска подстроки в строке алгоритмом Ахо-Корасик/main.py
from collections import deque
from typing import List

class AhoCorasick:
    def __init__(self, patterns: List[str]):
        self.num_nodes = 1  # стартовая вершина
        self.edges = [{}]  # список для хранения ребер
        self.fail = [-1]  # список для хранения функций неудачи
        self.output = [{}]  # список для хранения выходных значений

        # Строим автомат
        for pattern in patterns:
            self.add_pattern(pattern)

        # Строим функции неудачи с помощью BFS
        self.build_failure()

    def add_pattern(self, pattern: str):
        current_node = 0
        for char in pattern:
            if char not in self.edges[current_node]:
                self.edges[current_node][char] = self.num_nodes
                self.num_nodes += 1
                self.edges.append({})
                self.fail.append(-1)
                self.output.append({})
            current_node = self.edges[current_node][char]
        self.output[current_node][pattern] = True

    def build_failure(self):
        queue = deque()

        # Для первого уровня в дереве указываем неудачи на корень
        for char in range(256):  # Для всех символов ASCII
            char = chr(char)
            if char in self.edges[0]:
                self.fail[self.edges[0][char]] = 0
                queue.append(self.edges[0][char])
            else:
                self.edges[0][char] = 0

        # BFS для построения всех функций неудачи
        while queue:
            current_node = queue.popleft()

            for char in self.edges[current_node]:
                fail_state = self.fail[current_node]
                while char not in self.edges[fail_state]:
                    fail_state = self.fail[fail_state]
                self.fail[self.edges[current_node][char]] = self.edges[fail_state][char]
                self.output[self.edges[current_node][char]].update(self.output[self.fail[self.edges[current_node][char]]])
                queue.append(self.edges[current_node][char])

    def search(self, text: str) -> List[int]:
        current_node = 0
        result = []

        for i in range(len(text)):
            char = text[i]
            while char not in self.edges[current_node]:
                current_node = self.fail[current_node]
            current_node = self.edges[current_node][char]
            if self.output[current_node]:
                result.append(i)  # записываем индекс конца совпадения
        return result

def substring_search(text: str, pattern: str) -> List[int]:
    ac = AhoCorasick([pattern])
    return ac.search(text)

# Тестирование
if __name__ == "__main__":
    text = "stog siena igolka v stoge"
    pattern = "stog"
    result = substring_search(text, pattern)
    print(f"Найдено на позициях: {result}")

    # Тест с несколькими паттернами
    patterns = ["stog", "siena", "igolka"]
    ac = AhoCorasick(patterns)
    result_multiple = ac.search(text)
    print(f"Найдено на позициях: {result_multiple}")

File: Fundamentals_of_Software_Design/2_Algorithms/Поиска подстроки в строке алгоритмом Ахо-Корасик/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ALPHABET_SIZE 256  // Размер алфавита (для ASCII)

// Структура для хранения состояния автомата Ахо-Корасика
typedef struct AhoCorasick {
    int *fail;              // Массив функций неудачи
    int **edges;            // Массив ребер
    int *output;            // Массив выходных состояний (множество найденных паттернов)
    int num_nodes;          // Количество вершин
} AhoCorasick;

// Функция для создания нового автомата Ахо-Корасика
AhoCorasick* create_ac_automaton() {
    AhoCorasick *ac = (AhoCorasick *)malloc(sizeof(AhoCorasick));
    ac->num_nodes = 1;

    // Выделяем память для структур
    ac->edges = (int **)malloc(ALPHABET_SIZE * sizeof(int *));
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        ac->edges[i] = (int *)malloc(ALPHABET_SIZE * sizeof(int));
    }

    ac->fail = (int *)malloc(ALPHABET_SIZE * sizeof(int));
    ac->output = (int *)malloc(ALPHABET_SIZE * sizeof(int));
    return ac;
}

// Функция для добавления паттерна в автомат
void add_pattern(AhoCorasick *ac, const char *pattern) {
    int current_node = 0;
    for (int i = 0; pattern[i] != '\0'; i++) {
        unsigned char c = pattern[i];
        if (ac->edges[current_node][c] == -1) {
            ac->edges[current_node][c] = ac->num_nodes++;
        }
        current_node = ac->edges[current_node][c];
    }
    ac->output[current_node] = 1;  // Маркируем конец паттерна
}

// Функция для построения функций неудачи (BFS)
void build_failure(AhoCorasick *ac) {
    ac->fail[0] = 0;  // Для корня функция неудачи равна корню

    // Очередь для BFS
    int *queue = (int *)malloc(ac->num_nodes * sizeof(int));
    int front = 0, rear = 0;

    // Строим уровни для первого уровня
    for (int c = 0; c < ALPHABET_SIZE; c++) {
        if (ac->edges[0][c] != -1) {
            ac->fail[ac->edges[0][c]] = 0;
            queue[rear++] = ac->edges[0][c];
        }
    }

    // BFS для построения функции неудачи
    while (front < rear) {
        int current_node = queue[front++];
        for (int c = 0; c < ALPHABET_SIZE; c++) {
            if (ac->edges[current_node][c] != -1) {
                int fail_state = ac->fail[current_node];
                while (ac->edges[fail_state][c] == -1) {
                    fail_state = ac->fail[fail_state];
                }
                ac->fail[ac->edges[current_node][c]] = ac->edges[fail_state][c];
                queue[rear++] = ac->edges[current_node][c];
            }
        }
    }

    free(queue);
}

// Функция для поиска подстрок в строке
void search(AhoCorasick *ac, const char *text) {
    int current_node = 0;

    for (int i = 0; text[i] != '\0'; i++) {
        unsigned char c = text[i];

        // Следуем по дереву, если ребра не найдены, то идем по функции неудачи
        while (ac->edges[current_node][c] == -1) {
            current_node = ac->fail[current_node];
        }
        current_node = ac->edges[current_node][c];

        // Проверка на выходные состояния
        if (ac->output[current_node]) {
            printf("Подстрока найдена на позиции %d\n", i);
        }
    }
}

// Функция для инициализации и выполнения поиска
void substring_search(const char *text, const char *pattern) {
    AhoCorasick *ac = create_ac_automaton();

    // Добавляем паттерн
    add_pattern(ac, pattern);
    // Строим функции неудачи
    build_failure(ac);

    // Выполняем поиск в строке
    search(ac, text);

    // Освобождаем память
    free(ac->fail);
    free(ac->output);
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        free(ac->edges[i]);
    }
    free(ac->edges);
    free(ac);
}

int main() {
    const char *text = "stog siena igolka v stoge";
    const char *pattern = "stog";
    
    substring_search(text, pattern);  // Поиск одной подстроки

    return 0;
}
